internal/cmd/update.go

```diff
@@ -0,0 +1,466 @@
+		if strings.HasPrefix(a.Name, "coder-cli-"+ostype) {
+			assetURLStr = a.BrowserDownloadURL
+		}
+	}
+
+	if assetURLStr == "" {
+		return "", xerrors.Errorf("could not find release for ostype %s", ostype)
+	}
+
+	return assetURLStr, nil
+}
+
+func extractFromArchive(path string, archive []byte) ([]byte, error) {
+	contentType := http.DetectContentType(archive)
+	switch contentType {
+	case "application/zip":
+		return extractFromZip(path, archive)
+	case "application/x-gzip":
+		return extractFromTgz(path, archive)
+	default:
+		return nil, xerrors.Errorf("unknown archive type: %s", contentType)
+	}
+}
+
+func extractFromZip(path string, archive []byte) ([]byte, error) {
+	zipReader, err := zip.NewReader(bytes.NewReader(archive), int64(len(archive)))
+	if err != nil {
+		return nil, xerrors.Errorf("failed to open zip archive")
+	}
+
+	var zf *zip.File
+	for _, f := range zipReader.File {
+		if f.Name == path {
+			zf = f
+			break
+		}
+	}
+	if zf == nil {
+		return nil, xerrors.Errorf("could not find path %q in zip archive", path)
+	}
+
+	rc, err := zf.Open()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to extract path %q from archive", path)
+	}
+	defer rc.Close()
+
+	var b bytes.Buffer
+	bw := bufio.NewWriter(&b)
+	if _, err := io.Copy(bw, rc); err != nil {
+		return nil, xerrors.Errorf("failed to copy path %q to from archive", path)
+	}
+	return b.Bytes(), nil
+}
+
+func extractFromTgz(path string, archive []byte) ([]byte, error) {
+	zr, err := gzip.NewReader(bytes.NewReader(archive))
+	if err != nil {
+		return nil, xerrors.Errorf("failed to gunzip archive")
+	}
+
+	tr := tar.NewReader(zr)
+
+	var b bytes.Buffer
+	bw := bufio.NewWriter(&b)
+	for {
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, xerrors.Errorf("failed to read tar archive: %w", err)
+		}
+		fi := hdr.FileInfo()
+		if fi.Name() == path && fi.Mode().IsRegular() {
+			_, _ = io.Copy(bw, tr)
+			break
+		}
+	}
+
+	return b.Bytes(), nil
+}
+
+// getCoderConfigURL reads the currently configured coder URL, returning an empty string if not configured.
+func getCoderConfigURL() (*url.URL, error) {
+	urlString, err := config.URL.Read()
+	if err != nil {
+		return nil, err
+	}
+	configuredURL, err := url.Parse(strings.TrimSpace(urlString))
+	if err != nil {
+		return nil, err
+	}
+	return configuredURL, nil
+}
+
+// XXX: coder.Client requires an API key, but we may not be logged into the coder instance for which we
+// want to determine the version. We don't need an API key to hit /api/private/version though.
+func getAPIVersionUnauthed(client getter, baseURL url.URL) (*semver.Version, error) {
+	baseURL.Path = path.Join(baseURL.Path, "/api/private/version")
+	resp, err := client.Get(baseURL.String())
+	if err != nil {
+		return nil, xerrors.Errorf("get %s: %w", baseURL.String(), err)
+	}
+	defer resp.Body.Close()
+
+	ver := struct {
+		Version string `json:"version"`
+	}{}
+
+	body, err := ioutil.ReadAll(resp.Body)
```

Comment:
Ah, we should consider increasing our version requirement for coder-cli to 1.16 so that we can use io.ReadAll instead (ioutil stuff is now deprecated). No rush to do that, though, we can do it after 1.17 is released (should be soon)